### PR TITLE
ARROW-3398: [Rust] Update existing Builder to use MutableBuffer internally

### PR DESCRIPTION
**Pull request by paddyhoran**
I also updated a lint issue in buffer.rs that was caught by nightly rustfmt.  It's the same issue that was addressed in #2658 so I don't know how it got reverted.

There may be other lint issues that I will address as CI flags them (I hope not) as locally I am running into [this issue](https://github.com/rust-lang-nursery/rustfmt/issues/2916) after running `rustup update`

cc @kszucs @sunchao @andygrove

### CODE REVIEW

**Comment by paddyhoran** [comment]
The nightly rustfmt fix I mentioned above seems to cause current stable rustfmt to fail CI so I reset the branch.

File: rust/src/builder.rs [hunk 1]
```diff
@@ -15,158 +15,97 @@
...
-        let sz = mem::size_of::<T>();
-        unsafe {
-            libc::memcpy(
-                self.data.offset(self.len() as isize) as *mut libc::c_void,
-                slice.as_ptr() as *const libc::c_void,
-                slice.len() * sz,
-            );
-        }
-        self.len += slice.len();
+    /// Push a slice of type T, growing the internal buffer as needed
```
Comment (sunchao) [inline rust/src/builder.rs]: nit: add an empty line above.

File: rust/src/builder.rs [hunk 2]
```diff
@@ -15,158 +15,97 @@
...
-            libc::memcpy(
-                new_buffer as *mut libc::c_void,
-                old_buffer as *const libc::c_void,
-                self.len * sz,
-            );
-        }
-        self.capacity = new_capacity;
-        self.data = new_buffer as *mut T;
-        free_aligned(old_buffer as *const u8);
+    /// Grow the buffer to the new capacity of size n (where n is the number of elements of type T)
```
Comment (sunchao) [inline rust/src/builder.rs]: nit: add backtick to the parameters (i.e., `n` and `T`).

**Comment by sunchao** [comment]
Overall LGTM - just a few minor nits. More curious on how this is going to be used to build `ArrayData` and `Array`.

**Review by sunchao — APPROVED** [review]

File: rust/src/builder.rs [hunk 3]
```diff
@@ -15,158 +15,97 @@
...
-where
-    T: ArrowPrimitiveType,
-{
-    fn drop(&mut self) {
-        if !self.data.is_null() {
-            free_aligned(self.data as *const u8);
+    /// Writes a byte slice to the underlying buffer and updates the `len`, i.e. the number array
+    /// elements in the builder.  Also, converts the `io::Result` required by the `Write` trait
+    /// to the Arrow `Result` type.
+    fn write_bytes(&mut self, bytes: &[u8], len_added: i64) -> Result<()> {
```
Comment (kszucs) [inline rust/src/builder.rs]: Are there tests for it?

**Comment by paddyhoran** [comment]
@kszucs @sunchao thanks for the reviews, traveling at the moment. Will respond when I return.

File: rust/src/builder.rs (hunk 1 above)
Comment (paddyhoran) [reply rust/src/builder.rs]: Done

File: rust/src/builder.rs (hunk 3 above)
Comment (paddyhoran) [reply rust/src/builder.rs]: Added some.

File: rust/src/builder.rs (hunk 2 above)
Comment (paddyhoran) [reply rust/src/builder.rs]: Can you explain please @sunchao , I'm not quite understanding this one?

**Comment by paddyhoran** [comment]
@sunchao once this is merged I plan to take another look at [ARROW-3347](https://issues.apache.org/jira/browse/ARROW-3347?filter=-1) and use `BufferBuilder` internally for the values array and null bitmap.  

Are you still working on [ARROW-3089](https://issues.apache.org/jira/browse/ARROW-3089?filter=12343557)?  I don't want to step on your toes.  That JIRA (3089) could probably be broken down into several for each of the different builder types as it is it's scope is very broad.

**Comment by paddyhoran** [comment]
> Yes it will be better to break it down into sub tasks - maybe you can also move the ARROW-3347 under ARROW-3089? I can also pick up some tasks myself - I have some partial code on ARROW-3089 but nothing near complete right now, so feel free to create tasks and assign to yourself.

How can I "move it under"?  I mentioned it was a sub-task in the description but I can seem to officially make it a sub-task...

**Comment by sunchao** [comment]
In the JIRA page, you can click "More" and then "Convert to sub-task", in which you enter ARROW-3089.

**Comment by paddyhoran** [comment]
Ahh, thank you.  I was looking in "edit".  Done

**Comment by paddyhoran** [comment]
@kszucs the CI error is unrelated, so I think this is ready.  Thanks

**Comment by kszucs** [comment]
Great, merging! Thanks @paddyhoran !

**Pull request closed without merging**